File: app/src/main/java/moe/lpj/kandroid/viewmodel/MissionViewModel.java
package moe.lpj.kandroid.viewmodel;

import android.databinding.BaseObservable;
import android.databinding.Bindable;

import org.apache.commons.lang3.time.DateFormatUtils;
import org.apache.commons.lang3.time.DurationFormatUtils;

import java.util.Date;

import kandroid.observer.kcsapi.api_get_member;
import kandroid.observer.kcsapi.api_port;

public class MissionViewModel extends BaseObservable {

    private MissionViewModel() {
        api_port.port.INSTANCE.addListener(UtilsKt.getMissionViewModelUpdater());
        api_get_member.deck.INSTANCE.addListener(UtilsKt.getMissionViewModelUpdater());
    }

    private final static class Holder {
        public final static MissionViewModel Instance = new MissionViewModel();
    }

    public static MissionViewModel getInstance() {
        return Holder.Instance;
    }

    public boolean isMission2Begin;
    public String mission2Name;
    public Date mission2Time;

    public boolean isMission3Begin;
    public String mission3Name;
    public Date mission3Time;

    public boolean isMission4Begin;
    public String mission4Name;
    public Date mission4Time;

    @Bindable
    public String getMission2Name() {
        if (isMission2Begin) {
            return mission2Name;
        }
        return "null";
    }

    @Bindable
    public String getMission2Time() {
        if (isMission2Begin) {
            return DateFormatUtils.format(mission2Time, "HH:mm:ss");
        }
        return "null";
    }

    @Bindable
    public String getMission2LeftTime() {
        if (isMission2Begin) {
            return DurationFormatUtils.formatPeriod(new Date().getTime(), mission2Time.getTime(), "HH:mm:ss");
        }
        return "null";
    }

    @Bindable
    public String getMission3Name() {
        if (isMission3Begin) {
            return mission3Name;
        }
        return "null";
    }

    @Bindable
    public String getMission3Time() {
        if (isMission3Begin) {
            return DateFormatUtils.format(mission3Time, "HH:mm:ss");
        }
        return "null";
    }

    @Bindable
    public String getMission3LeftTime() {
        if (isMission3Begin) {
            return DurationFormatUtils.formatPeriod(new Date().getTime(), mission2Time.getTime(), "HH:mm:ss");
        }
        return "null";
    }

    @Bindable
    public String getMission4Name() {
        if (isMission4Begin) {
            return mission2Name;
        }
        return "null";
    }

    @Bindable
    public String getMission4Time() {
        if (isMission4Begin) {
            return DateFormatUtils.format(mission2Time, "HH:mm:ss");
        }
        return "null";
    }

    @Bindable
    public String getMission4LeftTime() {
        if (isMission4Begin) {
            return DurationFormatUtils.formatPeriod(new Date().getTime(), mission2Time.getTime(), "HH:mm:ss");
        }
        return "null";
    }
}

File: app/src/main/java/moe/lpj/kandroid/viewmodel/MissionViewModel.kt
package moe.lpj.kandroid.viewmodel

import android.databinding.ObservableArrayList
import kandroid.data.KCDatabase
import kandroid.observer.kcsapi.api_get_member
import kandroid.observer.kcsapi.api_port
import moe.lpj.kandroid.utils.NotificationUtils
import org.apache.commons.lang3.time.DateFormatUtils
import org.apache.commons.lang3.time.DurationFormatUtils
import java.util.*

object MissionViewModel {

    val NULL = Date() to "null"

    val missions = ObservableArrayList<Pair<Date, String>>()
    val leftTime = ObservableArrayList<String>()

    fun name(some: Pair<Date, String>): String {
        if (some !== NULL) {
            return some.second
        } else return "null"
    }

    fun time(some: Pair<Date, String>): String {
        if (some !== NULL) {
            return DateFormatUtils.format(some.first, "HH:mm:ss")
        } else return "null"
    }

    fun updateLeftTime() {
        for (i in 0..2) {
            val pair = missions[i]
            if (pair !== NULL) {
                val now = Date()
                if (pair.first.after(now)) {
                    leftTime.add(i, DurationFormatUtils.formatPeriod(now.time, pair.first.time, "HH:mm:ss"))
                } else leftTime.add(i, "00:00:00")
            } else leftTime.add(i, "null")
        }
    }

    val missionViewModelUpdater: () -> Unit = {
        val masterMissionData = KCDatabase.master.masterMissionData
        val fleets = KCDatabase.fleets

        //更新远征数据
        for (i in 0..2) {
            try {
                val fleet = fleets[i + 2]
                if (fleet != null && fleet.expeditionState != 0) {
                    missions.add(i, fleet.expeditionTime to (masterMissionData[fleet.expeditionDestination]?.name ?: "null"))
                } else {
                    missions.add(i, NULL)
                }
                log.info("更新第 ${i + 2} 舰队远征数据")
            } catch (e: Exception) {
                log.error("更新ViewModel失败：第 ${i + 2} 舰队远征数据 -> $e", e)
            }
        }

        //远征通知
        NotificationUtils.cancelMissionNotifications()
        NotificationUtils.registerMissionNotifications()
    }

    init {
        missions.add(NULL)
        missions.add(NULL)
        missions.add(NULL)
        api_port.port.addListener(missionViewModelUpdater)
        api_get_member.deck.addListener(missionViewModelUpdater)
    }
}